packages/task/src/browser/task-configurations.ts

```diff
@@ -0,0 +1,257 @@
+
+import { inject, injectable } from 'inversify';
+import { TaskOptions } from '../common/task-protocol';
+import { ILogger, Disposable, DisposableCollection } from '@theia/core/lib/common/';
+import URI from "@theia/core/lib/common/uri";
+import { FileSystemWatcherServer, DidFilesChangedParams, FileChange } from '@theia/filesystem/lib/common/filesystem-watcher-protocol';
+import { FileSystem } from '@theia/filesystem/lib/common';
+import * as jsoncparser from 'jsonc-parser';
+import { ParseError } from 'jsonc-parser';
+
+export interface TaskConfigurationClient {
+    /**
+     * A task has been added or removed from the known task configurations
+     * @returns an array of strings, each one being a task label
+     */
+    taskConfigurationChanged: (event: string[]) => void;
+}
+
+/**
+ * Watches potentially two tasks.json files: one in the user's home directory (in $home/.theia/ ),
+ * and  another one in the current workspace ( in $workspace/.theia/ )
+ */
+@injectable()
+export class TaskConfigurations implements Disposable {
+
+    /* Task client to send notifications to.  */
+    protected readonly toDispose = new DisposableCollection();
+
+    // Map where all the parsed tasks, from tasks.json files, are saved
+    protected tasksMap = new Map<string, TaskOptions>();
+
+    // the tasks.json config file(s) we are watching
+    protected watchedConfigFileUris: string[] = [];
+
+    /** last directory element under which we look for task config */
+    protected readonly TASKFILEPATH = '.theia';
+    /** task configuration file name */
+    protected readonly TASKFILE = 'tasks.json';
+
+    protected client: TaskConfigurationClient | undefined = undefined;
+
+    constructor(
+        @inject(ILogger) protected readonly logger: ILogger,
+        @inject(FileSystemWatcherServer) protected readonly watcherServer: FileSystemWatcherServer,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem
+    ) {
+        this.toDispose.push(watcherServer);
+        watcherServer.setClient({
+            onDidFilesChanged: changes =>
+                this.onDidTaskFileChange(changes)
+                    .then(() => {
+                        if (this.client !== undefined) {
+                            this.client.taskConfigurationChanged(this.getTaskLabels());
+                        }
+                    })
+                    .catch(err => {
+                        this.logger.error(err);
+                    })
+        });
+
+        this.toDispose.push(Disposable.create(() => {
+            this.tasksMap.clear();
+            this.client = undefined;
+        }));
+    }
+
+    setClient(client: TaskConfigurationClient) {
+        this.client = client;
+    }
+
+    dispose() {
+        this.toDispose.dispose();
+    }
+
+    watchWorkspaceConfiguration(wsRootUri: string): Promise<void> {
+        return this.watchConfigurationFile(wsRootUri);
+    }
+
+    protected watchConfigurationFile(wsRootUri: string): Promise<void> {
```

Comment:
you should try async/await. It would help understanding the code.

Reply:
ok, I am giving this a try: I will attempt to use async/await in this file.

Reply:
Done (will push updated version soon).